dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -309,4 +348,51 @@ void MPPTaskManager::releaseThreadsFromScheduler(const int needed_threads)
     scheduler->releaseThreadsThenSchedule(needed_threads, *this);
 }
 
+void MPPTaskManager::monitorMPPTasks()
+{
+    std::unique_lock lock(mu, std::defer_lock);
+    while (true)
+    {
+        lock.lock();
+
+        // Check MPPTasks every 5 minutes
+        monitor_cv.wait_for(lock, std::chrono::seconds(300));
+
+        if (is_shutdown)
+        {
+            lock.unlock();
+            return;
+        }
+
+        checkMPPTasks(monitored_tasks, log);
+        lock.unlock();
+    }
+}
+
+void MPPTaskManager::addMonitoredTask(const String & task_unique_id)
+{
+    std::lock_guard lock(mu);
+    auto iter = monitored_tasks.find(task_unique_id);
+    if (iter != monitored_tasks.end())
+    {
+        LOG_ERROR(log, "task {} is repeatedly added to be monitored which is not an expected behavior!");
+        return;
+    }
+
+    monitored_tasks.insert(std::make_pair(task_unique_id, Stopwatch()));
+}
+
+void MPPTaskManager::removeMonitoredTask(const String & task_unique_id)
+{
+    std::lock_guard lock(mu);
+    auto iter = monitored_tasks.find(task_unique_id);
+    if (iter == monitored_tasks.end())
+    {
+        LOG_ERROR(log, "Unexpected behavior! task {} is not found in monitored_task.");
```

Comment:
Do we consider change to Info level? It would allow removeMonitorerTask been called multiple times for the same task.